File: backend/src/providers/notificationProvider.ts
import { prisma } from '../lib/prisma.js';

/**
 * Data Access Layer — Notifications
 */
export const notificationProvider = {
  /** Create a notification for a specific user. */
  async create(data: {
    userId: bigint;
    type: string;
    title: string;
    body: string;
    refId?: bigint;
  }) {
    return prisma.notifications.create({
      data: {
        user_id: data.userId,
        type: data.type,
        title: data.title,
        body: data.body,
        ref_id: data.refId,
      },
    });
  },

  /** Return the 50 most recent notifications for a user, newest first. */
  async findByUserId(userId: bigint) {
    return prisma.notifications.findMany({
      where: { user_id: userId, deleted: false },
      orderBy: { created: 'desc' },
      take: 50,
    });
  },

  /** Find a single notification by id. */
  async findById(id: bigint) {
    return prisma.notifications.findFirst({
      where: { id, deleted: false },
    });
  },

  /** Mark a notification as read. */
  async markRead(id: bigint, readAt: Date) {
    return prisma.notifications.update({
      where: { id },
      data: { is_read: true, read_at: readAt },
    });
  },

  /**
   * Find all active staff/admin users (employee, chef, admin roles).
   * Used to fan-out notifications when a customer posts a comment.
   */
  async findAllStaffAdminIds(): Promise<bigint[]> {
    const users = await prisma.user.findMany({
      where: {
        deleted: false,
        active: true,
        roles: {
          some: {
            deleted: false,
            active: true,
            role: {
              deleted: false,
              active: true,
              name: { in: ['admin', 'employee', 'chef'] },
            },
          },
        },
      },
      select: { id: true },
    });
    return users.map((u) => u.id);
  },
};

File: backend/app/providers/notification_provider.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update
from typing import Optional, List
from app.models.notification import Notification

class NotificationProvider:
    async def find_by_user(self, db: AsyncSession, user_id: int, limit: int = 50) -> List[Notification]:
        result = await db.execute(
            select(Notification)
            .where(Notification.user_id == user_id, Notification.deleted == False)
            .order_by(Notification.created.desc())
            .limit(limit)
        )
        return list(result.scalars().all())

    async def create(self, db: AsyncSession, user_id: int, type: str, title: str, body: str, ref_id: Optional[int] = None) -> Notification:
        notif = Notification(user_id=user_id, type=type, title=title, body=body, ref_id=ref_id)
        db.add(notif)
        await db.commit()
        await db.refresh(notif)
        return notif

    async def mark_read(self, db: AsyncSession, notif_id: int, user_id: int) -> Optional[Notification]:
        from datetime import datetime, timezone
        await db.execute(
            update(Notification).where(Notification.id == notif_id, Notification.user_id == user_id).values(is_read=True, read_at=datetime.now(timezone.utc))
        )
        await db.commit()
        result = await db.execute(select(Notification).where(Notification.id == notif_id))
        return result.scalar_one_or_none()

notification_provider = NotificationProvider()
